### DISCUSSION TITLE
TypeIsFunction#typeIs and Lombok

### DISCUSSION THREAD

**Discussion by dykov** [pmd/pmd#3708] [Q&A] [answered]
Hi,
Can I check if a class uses `org.slf4j.Logger` if it annotated with `lombok.extern.slf4j.Slf4j`?
As I checked, `TypeIsFunction#typeIs` doesn't work for the case:
```java
import lombok.extern.slf4j.Slf4j;
@Slf4j
public final class Foo {
    void method() {
        log.error("message");
    }
}
```
XPath rule:
```xml
//Name[pmd-java:typeIs('org.slf4j.Logger')]
```

Thank you

**Comment by jsotuyod** [comment, marked as answer]
I'm unsure what line you would expect to flag… Is it this one `log.error("message");`?

Any reason to not flag the usage of `@lombok.extern.slf4j.Slf4j` altogether?

**Reply by dykov** [reply]
Right, `log.error` invocation

**Reply by jsotuyod** [reply]
The thing is, the expression `log.error("message")` is actually of type `void`. Only `log` could potentially be resolved to `org.slf4j.Logger`, but PMD currently doesn't split / resolve the segments of a `Name` individually.

The best thing I can think of would be:

```xpath
//TypeDeclaration (: within a class … :)
  [Annotation[pmd-java:typeIs('lombok.extern.slf4j.Slf4j')]] (: annotated with @Slf4j… :)
  //Name[matches(@Image, '(this.)?log(\..*)?')] (: … mark usages of the "log" field :)
```

**Reply by dykov** [reply]
Thanks a lot!
